pkg/activator/prober.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package activator
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"go.uber.org/zap"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	maxRetry              = 60
+	defaultPeriodSeconds  = int32(1 * time.Second)
+	defaultTimeoutSeconds = int32(1 * time.Second)
+)
+
+func verifyRevisionRoutability(revision *v1alpha1.Revision, endpoint *Endpoint, logger *zap.SugaredLogger) {
+	// Proceed only if user has a HTTPGet readinessProbe defined
+	if revision.Spec.Container.ReadinessProbe == nil ||
+		revision.Spec.Container.ReadinessProbe.HTTPGet == nil {
+		endpoint.Verified = Unknown
+		return
+	}
+
+	endpoint.Verified = Fail
+	probe := createHttpGetProbe(revision, *endpoint)
+
+	// Number of seconds after the readiness probes are initiated
+	time.Sleep(time.Second * int32ToDuration(probe.InitialDelaySeconds))
+
+	retryCount := 1
+	retryInterval := time.Second * int32ToDuration(probe.PeriodSeconds)
+
+	for retryCount = 1; retryCount < maxRetry; retryCount++ {
+		ready, err := checkHttpGetProbe(probe, logger)
+		if err != nil {
+			logger.Errorf("error checking probe", zap.String("error", err.Error()))
+		}
+
+		if ready {
+			endpoint.Verified = Pass
+			break
+		}
+
+		// How often (in seconds) to perform the probe
+		time.Sleep(retryInterval)
+	}
+
+	logger.Infof("took %d probe retries for readiness", retryCount, zap.Any("endpoint", endpoint))
+}
+
+// Function creates HTTP readiness probe for revision
+func createHttpGetProbe(revision *v1alpha1.Revision, endpoint Endpoint) *v1.Probe {
```

Comment:
You're transforming a revision, endpoint to a probe. Then in the subsequent steps you're converting the probe to an http get call/request.

You can go straight to a (revision, endpoint) -> http.Request/call

Reply:
Fair. This approach is a remnant of the initial effort to tackle both HttpGet and TCPSocket based readiness probes. Support for TCPSocket probe is currently blocked by issue #1241. More details in this [issue comment](https://github.com/knative/serving/issues/1448#issuecomment-405684894). 

Will update.